ArduPlane/mode_autoland.cpp

```diff
@@ -0,0 +1,119 @@
+#include "mode.h"
+#include "Plane.h"
+#include <GCS_MAVLink/GCS.h>
+
+/*
+  mode AutoLand parameters
+ */
+const AP_Param::GroupInfo ModeAutoLand::var_info[] = {
+    // @Param: WP_ALT
+    // @DisplayName: Final approach WP altitude
+    // @Description: This is the target altitude above HOME for final approach waypoint
+    // @Range: 0 200
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_ALT", 1, ModeAutoLand, final_wp_alt, 55),
+
+    // @Param: WP_DIST
+    // @DisplayName: Final approach WP distance
+    // @Description: This is the distance from Home that the final approach waypoint is set. The waypoint point will be in the opposite direction of takeoff (the direction the plane is facing when the plane sets its takeoff heading)
+    // @Range: 0 700
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_DIST", 2, ModeAutoLand, final_wp_dist, 400),
+
+    AP_GROUPEND
+};
+
+ModeAutoLand::ModeAutoLand() :
+    Mode()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+bool ModeAutoLand::_enter()
+{     
+    //if do_land_start exists, jump to it
+    if( (plane.mission.contains_item(MAV_CMD_DO_LAND_START)) && (plane.arming.is_armed())) {
+        if (plane.have_position && plane.mission.jump_to_landing_sequence(plane.current_loc)) {
+            // switch to AUTO
+            plane.mission.set_force_resume(true);
+            if (plane.set_mode(plane.mode_auto, ModeReason::FIXED_WING_AUTOLAND)) {
+                 return true;
+            }
+        }
+    }
+             
+    //also must be flying to enter
+    if (!plane.is_flying()) { 
+        return false;
+    }
+    
+    //takeoff direction must be set and must not be a quadplane, otherwise since flying switch to RTL so this can be used as FS action
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "AutoLand is fixed wing only,changing to RTL"); 
+        plane.set_mode(plane.mode_rtl, ModeReason::FIXED_WING_AUTOLAND);
```

Comment:
there are 3 reasons to fail...two result in switch to RTL...its a lot less code to do the switch in the enter rather than hack around system.cpp and change the architecture of set_mode, IMHO

Reply:
Well, here your assuming that the user wants RTL. They might prefer to do something else. If you refuse the mode enter they can decide. Maybe they want QRTL or QLAND instead of RTL. These eventualities result in us having to add more Q_OPTIONS in the future, better to just reject the mode change.

Reply:
okay no mode changes in enter